### PR TITLE
Af packet new modes v1.2

### PR DESCRIPTION
**Pull request by regit**
A rebase of #1565.

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/85
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/83

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1054,8 +1054,8 @@
             [enable_af_packet="no"],
             [[#include <sys/socket.h>
               #include <linux/if_packet.h>]])
-        AC_CHECK_DECL([PACKET_FANOUT],
```
Comment (inliniac) [inline configure.ac]: Will this disable our FANOUT support for any kernel not supporting FANOUT_QM?

Comment (regit) [inline configure.ac]: No, see https://github.com/regit/suricata/blob/af-packet-new-modes-v1.2/src/source-af-packet.h#L27, we are forcing the declaration if the headers do not have the define.

Comment (inliniac) [inline configure.ac]: I see. So really the whole test is sort of useless?

Comment (regit) [inline configure.ac]: Hi,
Yes, no functional change and for now no impact that can be foreseen.
Le 21 juil. 2015 11:38 PM, Victor Julien <notifications@github.com> a écrit :In configure.ac:
> @@ -1054,8 +1054,8 @@
>              [enable_af_packet="no"],
>              [[#include <sys/socket.h>
>                #include <linux/if_packet.h>]])
> -        AC_CHECK_DECL([PACKET_FANOUT],

I see. So really the whole test is sort of useless?

**Comment by inliniac** [comment]
I think some of the new options are quite confusing. How I understand them:

* cluster_random: random distribution of packets over the sockets? In other words, useless for any use case where stateful protocols like TCP are used as timing issues will break the order in which we process the packets.

* cluster_qm: balances by rss queue, so the balancing is done by the nic and not the kernel

* cluster_rollover: I have no idea what this does

* rollover option: modifies all existing cluster modes to spill over packets from one socket to the next, so that we don't drop them. However due to the timing issues also present in 'cluster_random', this will lead to massive TCP (and other stateful protocol) tracking issues. This is what I see in my replay test as well (cluster_flow+rollover). Drops are way down, but so are the alerts. So despite inspecting more packets we're missing more alerts.

Can you confirm and fill in the blanks?

**Comment by regit** [comment]
Hi,

- cluster_random: useless for us indeed
- cluster_qm: your analysis is perfect
- cluster_rollover: send packet to first socket, when ring is full send them to next socket with room available (I think it is useless too)
- rollover: analysis correct. Strange you see less alerts. We may find something to fix or improve here.

**Comment by inliniac** [comment]
Is the rollover option enforced per packet or is a flow 'migrated' to a new socket completely? Assuming the first.

It does make sense to me that alert rates would be lower. My test measures drop rates, and the base line is about 30-50% drop (depending on runmode and options). This means that many flows will be affected.

What will happen is that for many flows multiple threads will be processing the packets at approximately the same time. Due to threading timing uncertainty, we will see many cases where ordering is off. For example process a SYN/ACK before the SYN. All kinds of issues will happen: packets considered to be out of window, bad RST's, bad ACK's, etc, etc.

In my test, I see a slight reduction in drop rates, but almost 50% less alerts.

**Comment by regit** [comment]
Yes, rollover is per-packet.

Regarding your alert rate, you are testing a very bad case (30% drop). I understand that in that case it can just help to get thing a bit worse. I will try to check with lower drop rate that appear to be fixed by the setup.

**Comment by inliniac** [comment]
Rerunning my tests at lower speeds as well.

I think it may be good to hide all the 'useless' options from the yaml and just point to a wiki page where the various options and their effects are explained. There may be a case for each of them, but most users shouldn't even consider cluster_random and cluster_rollover for example.

**Comment by regit** [comment]
Yes, make sense.

**Comment by inliniac** [comment]
Ok cool, I'll be merging this PR today. Such changes can be incremental.

**Comment by inliniac** [comment]
rolloverstats

Ok there seems to be a slight advantage when drop rate is close to 0, but in all other cases I see massive alert reduction, even when drop rates align. The red line is key here.

Left Y-axis is alert count. Right Y-axis drop rate in %. X-axis replay speed in Gbps.

**Comment by inliniac** [comment]
Btw drop rate is calculated by taking number of sent packets (known quantity) and the known processed packets (stats.log). The drop stats suri reports are ignored.

**Comment by regit** [comment]
Really interesting. What kind of alerts are generated (you've got heavy number here) ? Do you the stats for the metadata ?

**Comment by inliniac** [comment]
This based on etpro-all.rules from a while back, on a 118GiB pcap with much HTTP traffic. I'm not yet tracking HTTP sessions/logs or TCP session in my testing. Will see if I can add that.

What kind of metadata would you be interested in?

**Comment by regit** [comment]
Basically, I would like to see if we have the same ratio in metadata generation. HTTP for example would be nice.

**Comment by inliniac** [comment]
rolloverstats2
Speed vs drop rate vs http tx log count. Pretty clear that the rollover leads to massive issues.

**Comment by inliniac** [comment]
I tested the same branch with the rollover opt disabled and then it behaves like master. So it's not caused by some other diff between the branches.

**Comment by poona** [comment]
Is there a difference between cluster_qm and cluster_cpu?

My understanding was that cluster_cpu depended on rss, and hence the nic to do the balancing, which is the same thing as cluster_qm?

**Comment by regit** [comment]
@poona: It is sort of the same but it is possible to pin some flow to a given qm in the network card. So in that case that can be really different.

**Comment by pevma** [comment]
@inliniac  - do you get the exact same packet stats with both runs using the same pcap ? (ex kernel.packets/bytes..)

**Comment by inliniac** [comment]
My tool isn't tracking the stats.log yet, it uses the stats dumped at shutdown instead. From there it takes pkts seen and pkts dropped. Since I know exactly how many packets were sent (using the same pcap each time) I can with certainty tell what the drop rate is. Alerts and http counts are simple 'wc -l' on the http.log and fast.log.

**Comment by pevma** [comment]
I did some live traffic tests - on one of our test boxes. I did not see such a big discrepancy in the runs with and without this PR. Could this be related to some particularity in a pcap replay technique?

**Comment by inliniac** [comment]
Merged through #1607. We'll continue to look at that rollover issue in the coming time.

**Pull request closed without merging**